commit 2db79c96baa1256b8fc2656596143da92fabd074
Author: Sergey Kolosov <skolosov@redhat.com>
Date:   Wed Apr 10 17:58:04 2024 +0200

    libsupport: Add xgetpeername
    
    The patch adds redirections for getpeername.
    
    Reviewed-by: Arjun Shankar <arjun@redhat.com>
    (cherry picked from commit 6687a6e3f962759536a8019d31c68c1009ccd6eb)

diff --git a/support/Makefile b/support/Makefile
index 362a51f882787f2b..aa57207bdccc852d 100644
--- a/support/Makefile
+++ b/support/Makefile
@@ -131,6 +131,7 @@ libsupport-routines = \
   xfreopen \
   xftruncate \
   xgetline \
+  xgetpeername \
   xgetsockname \
   xlisten \
   xlseek \
diff --git a/support/xgetpeername.c b/support/xgetpeername.c
new file mode 100644
index 0000000000000000..6f448e456a1d9e1e
--- /dev/null
+++ b/support/xgetpeername.c
@@ -0,0 +1,30 @@
+/* getpeername with error checking.
+   Copyright (C) 2024 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include <support/xsocket.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <support/check.h>
+
+void
+xgetpeername (int fd, struct sockaddr *sa, socklen_t *plen)
+{
+  if (getpeername (fd, sa, plen) != 0)
+    FAIL_EXIT1 ("getpeername (%d): %m", fd);
+}
diff --git a/support/xsocket.h b/support/xsocket.h
index 3e4410354676cb2a..4ac0e1f5ffe35dc2 100644
--- a/support/xsocket.h
+++ b/support/xsocket.h
@@ -26,6 +26,7 @@
 int xsocket (int, int, int);
 void xsetsockopt (int, int, int, const void *, socklen_t);
 void xgetsockname (int, struct sockaddr *, socklen_t *);
+void xgetpeername (int, struct sockaddr *, socklen_t *);
 void xconnect (int, const struct sockaddr *, socklen_t);
 void xbind (int, const struct sockaddr *, socklen_t);
 void xlisten (int, int);
